### DISCUSSION TITLE
MESA-LLM

### DISCUSSION THREAD

**Discussion by Spartan-71** [mesa/mesa#2679] [Future of Mesa]
Hello everyone! I am interested in working on the mesa-llm project.
I have created this new discussion so that all the comments/discussions are in one place.

**Comment by Spartan-71** [comment]
As mentioned in the Project wiki mesa-llm should have reusable components like planning, memory and reasoning modules.

Can we use Langchain, Crewai, or any other agentic framework to integrate LLM into Mesa, or should we just refer to how these frameworks are built and then build relevant components from scratch?

**Reply by wang-boyu** [reply]
At this point we're open to ideas on whether to use third party frameworks or not

**Comment by WingchunSiu** [comment]
I was thinking the same thing. I think existing frameworks like LangChain and CrewAI are useful for structured workflows but may be too rigid for Mesa-LLM’s modular needs. They encapsulate the planning, memory, and reasoning a bit too much, which simplifies development but may limits deep customization for abm.

A more flexible alternative I could think of is LangGraph, which allows graph-based decision flows. For example adding an LLM-based supervisor to oversee agent reasoning can help maintain adaptability, but ultimately we need highly customizable components tailored to Mesa’s ecosystem. So rather than forcing an external framework it makes more sense to design modular and extensible agent structures I guess.

Curious to hear thoughts—do we see areas where existing frameworks would help, or is customization unavoidable for integration?

**Reply by colinfrisch** [reply]
Hey, I've worked with the HuggingFace's smolagent framework and it could be a reasonable start. Since the library is very easy of use, it could really simplify the coding compared to doing some function calling using sthg like the OpenAI API directly. The other advantage is reasoning (pre-implemented). In my opinion, it would not be incompatible with a LangGraph based structure. What do you think ?


<pre lang="python">from mesa import Agent

class LLMAgent(Agent):
    def __init__(self, unique_id, model, smol_agent):
        super().__init__(unique_id, model)
        self.smol_agent = smol_agent

    def step(self):
        task = "Describe the current environment state and --take action with available function tools--"
        response = self.smol_agent.run(task)


class SimulationModel(Model):
    def __init__(self):
        self.smol_agent = CodeAgent(tools=[TOOL_LIST], model=HfApiModel(model_id="MODEL_TO_USE"))

        for i in range(number_of_agents):
            agent = LLMAgent(i, self, self.smol_agent)
            self.schedule.add(agent)  </pre>

**Reply by piratevector** [reply]
@colinfrisch I think this snippet is using Mesa 2.x

**Reply by colinfrisch** [reply]
> @colinfrisch I think this snippet is using Mesa 2.x

It's not a code that's currently implemented in mesa, just a suggestion for a feature :)

**Comment by Bhavishya-Chaturvedi** [comment]
Hello @Spartan-71 and @WingchunSiu, I'd like to collaborate on this project as well. I haven't found much discussion consolidated in one place. Are there any recent developments or a central discussion area I might have missed?

**Reply by WingchunSiu** [reply]
Hey, you can join Mesa's group discussion on https://summerofcode.withgoogle.com/programs/2025/organizations/project-mesa through the "chat" on the right.

**Reply by shubhambhar007** [reply]
Hi @WingchunSiu is there any ongoing discussion in github?
I am very excited to apply for this project in this upcoming GSoC and would really appreciate your guidance on the same.
Thanks!

**Reply by wang-boyu** [reply]
This is a good place to discuss Mesa-LLM. And of couse using matrix chat is also encouraged!

**Reply by Bhavishya-Chaturvedi** [reply]
Thanks a lot @shubhambhar007  @wang-boyu

**Comment by wang-boyu** [comment]
I have answeres some questions raised in a separate discussion about mesa-llm. Here is the link in case it's of interests to some others: https://github.com/projectmesa/mesa/discussions/2046#discussioncomment-12605328

**Comment by Bhavana-bhuvana** [comment]
Hi everyone,

I'm Bhavana S final year B.C.A student new to open source and really excited to start contributing. I’m currently exploring the MESA-LLM: Generative Agent-Based Modeling with LLM-Empowered Agents project for GSoC 2025 and preparing my proposal. I’ve started going through the Mesa documentation, some practical examples, and also the CONTRIBUTING.md file to understand how to get started the right way.

I have experience with Python, data analysis, and machine learning using Scikit-learn. I’m also exploring deep learning and LLMs, and have worked on frontend projects using Streamlit. Recently, I started learning React to expand my frontend skills.

I would really appreciate any guidance on how to study the Mesa framework and this project more effectively. If there are specific resources, issues, or tasks you'd recommend, I’d love to work on them. I’m flexible and open to learning, and can align my study path with your suggestions.

Thanks a lot, and I’m looking forward to contributing and learning more with the community!

**Comment by mawa-maria** [comment]
Hi @Spartan-71, and the Mesa team,

I'm Fatima-ezzahra, an AI/Computer Engineering student at ENSAM Casablanca, excited to apply for GSoC 2025 to work on Mesa-LLM. Here's how I plan to contribute:

### **Next Steps**
1. **First Contributions** (starting this week):
   - Tackle "good first issues" to understand Mesa's architecture
   - Improve documentation (e.g., add LLM integration examples)
   
2. **Technical Approach** (for the project):
   ```python
   class LLMAgent(Agent):
       def __init__(self, unique_id, model, llm_backend="huggingface"):
           super().__init__(unique_id, model)
           self.llm = HuggingFacePipeline()  # Default backend
           self.memory = ShortTermMemory()  # Customizable module
Focus on modularity (swap LLM backends easily)

Build debug tools (LLM decision logging)

Why I'm a Good Fit
Skills: Python, ML (IBM Data Science cert), basic Mesa knowledge

Projects: Built AI systems like [Calorie Tracker] (recommendation engine)

Availability: 56h/week dedicated to GSoC (May-Aug)

Questions
Which LLM backends are highest priority? (HuggingFace/OpenAI)

Are there example models I should study first?

I'll begin working through open issues and share progress here. Looking forward to your guidance!

Best,
Fatima-ezzahra
GitHub: [mawa-maria](https://github.com/mawa-maria)

**Reply by Spartan-71** [reply]
hey @mawa-maria, great to know that you are also interested in MESA-LLM project. 

> Which LLM backends are highest priority? (HuggingFace/OpenAI)

As mentioned in the [Project Wiki](https://github.com/projectmesa/mesa/wiki/Google-Summer-of-Code-2025), I think there is no such priority from the Mesa team for the backend; they need integrations with as much as LLMs as possible. To ensure the same, we can use [LiteLLM](https://docs.litellm.ai/docs/), which is a LLM Gateway / Proxy Server.

> Are there example models I should study first?

I recommend you go through the [introductory tutorial](https://mesa.readthedocs.io/latest/tutorials/0_first_model.html), which will help you in understanding "The Boltzmann Wealth Model" in detail, with the implementation.

**Comment by AlexandreFln** [comment]
Hi everyone, if it can be of any interest I tried to design an energy market simulation using LLM-powered agents with Mesa framework. This is a good start but it can largely be improved and I would be happy to hear your thoughts on it.

[Here](https://github.com/AlexandreFln/llm-abm-energymarkets) is the repo

**Reply by wang-boyu** [reply]
Thanks for sharing! We'll definitely look into it while working on mesa-llm in the following months.

**Comment by adarshkumar23** [comment]
Hi everyone! 👋

I'm Adarsh Kumar, a student at IIT Patna and an AI/ML Engineer with production experience at ValersAI, where I progressed from ML Intern to ML Engineer working with BERT, GPT-2, and LLM pipelines deployed on AWS.

I'm very interested in the Mesa-LLM project for GSoC 2026. My background is directly relevant:

- Hands-on experience with LLM APIs (OpenAI, HuggingFace Transformers)
- Production Python development with PyTorch and TensorFlow
- Experience with async programming and API integration
- Currently building a personal AI assistant as a side project

I've cloned the mesa-llm repo and started exploring the codebase. A few initial observations:

1. The library is still in early experimental state — I'd love to help stabilize the API and improve test coverage
2. I noticed the discussion about using LangChain/CrewAI vs building custom components — I lean toward the modular custom approach for better Mesa integration, but I'd love to hear the mentors' preferred direction
3. I'm planning to build a small example model this week to identify friction points in the current API

Questions for the mentors:
- Are there specific open issues in mesa-llm that would be good starter contributions?
- What's the preferred approach for handling API rate limiting and latency across multiple LLM-powered agents?
- Is there a target date for mesa 4.0.0, and how should mesa-llm prepare for the breaking changes?

Looking forward to contributing!

**Comment by adarshkumar23** [comment]
Hi @wang-boyu and Mesa-LLM maintainers!

I'm Adarsh Kumar, applying for the Mesa-LLM GSoC 2026 project. 
I've already contributed to Mesa with PR #229 (fix for speak_to bug, issue #158).

I have 2 years of professional experience as an ML Engineer working with LLM APIs (OpenAI, Anthropic, Gemini, Ollama), HuggingFace Transformers, prompt engineering, and async Python. I've built production agentic systems with reasoning and memory components — which aligns directly with Mesa-LLM's architecture.

I've cloned mesa-llm, run the existing examples, and identified a few areas I could start contributing to immediately:
- Improving test coverage for the reasoning and memory modules
- Adding documentation for the tool/function-calling workflow
- Exploring API latency optimization with async batching

Before I start my proposal draft, I'd appreciate guidance on:
1. What are the highest-priority open issues for mesa-llm right now?
2. Is there a preferred direction for the scientific paper-based example models?
3. Any specific friction points in the current API that you'd want addressed first?

Happy to pick up any open issues immediately. Looking forward to your feedback!

GitHub: https://github.com/adarshkumar23
